src/Umbraco.Infrastructure/PropertyEditors/MediaPicker3PropertyEditor.cs

```diff
@@ -383,5 +391,101 @@ public IEnumerable<ValidationResult> Validate(
+        {
+            private readonly ILocalizedTextService _localizedTextService;
+            private readonly IMediaTypeService _mediaTypeService;
+
+            public AllowedTypeValidator(ILocalizedTextService localizedTextService, IMediaTypeService mediaTypeService)
+            {
+                _localizedTextService = localizedTextService;
+                _mediaTypeService = mediaTypeService;
+            }
+
+
+            public IEnumerable<ValidationResult> Validate(
+                List<MediaWithCropsDto>? value,
+                MediaPicker3Configuration? configuration,
+                string? valueType,
+                PropertyValidationContext validationContext)
+            {
+                if (value is null || configuration is null)
+                {
+                    return [];
+                }
+
+                var allowedTypes = configuration.Filter?.Split(Constants.CharArrays.Comma, StringSplitOptions.RemoveEmptyEntries);
+
+                // No allowed types = all types are allowed
+                if (allowedTypes is null || allowedTypes.Length == 0)
+                {
+                    return [];
+                }
+
+                foreach (MediaWithCropsDto media in value)
+                {
+                    IMediaType? type = _mediaTypeService.Get(media.MediaTypeAlias);
+
+                    if (type is null || allowedTypes.Contains(type.Key.ToString()) is false)
+                    {
+                        return
+                        [
+                            new ValidationResult(
+                                _localizedTextService.Localize("validation", "invalidMediaType"),
+                                new[] { "invalidMediaType" })
+                        ];
+                    }
+                }
+
+                return [];
+            }
+        }
+
+        internal class StartNodeValidator : ITypedJsonValidator<List<MediaWithCropsDto>, MediaPicker3Configuration>
+        {
+            private readonly ILocalizedTextService _localizedTextService;
+            private readonly IMediaNavigationQueryService _mediaNavigationQueryService;
+
+            public StartNodeValidator(
+                ILocalizedTextService localizedTextService,
+                IMediaNavigationQueryService mediaNavigationQueryService)
+            {
+                _localizedTextService = localizedTextService;
+                _mediaNavigationQueryService = mediaNavigationQueryService;
+            }
+
+            public IEnumerable<ValidationResult> Validate(
+                List<MediaWithCropsDto>? value,
+                MediaPicker3Configuration? configuration,
+                string? valueType,
+                PropertyValidationContext validationContext)
+            {
+                if (value is null || configuration?.StartNodeId is null)
+                {
+                    return [];
+                }
+
+                foreach (MediaWithCropsDto media in value)
```

Comment:
Similarly, you could perhaps group by parent here and look up via ancestor or self?  That way you do the same test, but if you had several media from a single folder there would be less to loop through.

Reply:
This turned out to be quite a bit more tricky to optimize, but I think I've reduced it as much as you can, lookup/iteration wise, I've added additional tests to check it works correctly.

Reply:
Yes, just had a look, that was more effort that I thought when writing the comment, so apologies for that!  Looks good though with the update, and likely worth it if you have quite a number of picked media to validate.